### PR TITLE
Fix allreduce with inplace

### LINKED ISSUE #287: Perform MPI_ALLREDUCE operations in place when possible
Several all-to-all MPI reductions are using a temporary variable : 

https://github.com/xcompact3d/Incompact3d/blob/55652680be067db9336161b2a9c873744c637811/src/Case-TGV.f90#L298

However, MPI reductions can be performed in place to avoid temporary variables : 

https://github.com/xcompact3d/Incompact3d/blob/55652680be067db9336161b2a9c873744c637811/src/navier.f90#L1541

It would be a nice improvement to perform MPI all-to-all reductions in place when possible and to remove temporary variables, especially when their name is not clear (`temp1`, `temp2`, ...)

### PR DESCRIPTION
**Pull request by slaizet**
This is to deal with issue #287 

I have listed 186 ALLREDUCE operations in the codebase. 181 of them are now using MPI_IN_PLACE for clarity. Only 5 of them are not using MPI_IN_PLACE.

I have tested the codebase extensively but more testing is required.

### CODE REVIEW

File: src/acl_source.f90 [hunk 1]
```diff
@@ -520,7 +519,7 @@ subroutine Compute_Momentum_Source_Term_pointwise
       enddo
 
       alm_proj_time=MPI_WTIME()-t1
-      call MPI_ALLREDUCE(alm_proj_time,t1,1,real_type,MPI_SUM,MPI_COMM_WORLD,ierr)
+      call MPI_ALLREDUCE(MPI_IN_PLACE,alm_proj_time,1,real_type,MPI_SUM,MPI_COMM_WORLD,ierr)
```
Comment (mathrack) [inline src/acl_source.f90]: This modification is fixing a minor issue ? The code was printing the local value before. Now it is printing the global value obtained after reduction.

Comment (slaizet) [reply src/acl_source.f90]: possibly a typo on my side :-)

**Review by mathrack — APPROVED** [review]

File: src/acl_source.f90 (hunk 1 above)
Comment (mathrack) [reply src/acl_source.f90]: The other modifications are looking equivalent to the current code. This is probably the only one changing the behavior.

Comment (mathrack) [reply src/acl_source.f90]: Before the result of the reduction (`t1`) was unused

Comment (slaizet) [reply src/acl_source.f90]: I will have a look and fix the typo

Comment (mathrack) [reply src/acl_source.f90]: As a side remark, MPI_Wtime is returning a double precision number. If accuracy is needed on the timer, the single precision builds should not store the result in a variable defined with `real(mytype)` but inside a double precision variable. However, if there is no need for accuracy in the timer, it is ok to store it inside a variable `real(mytype)`

Comment (mathrack) [reply src/acl_source.f90]: I think the new version is better, no need to fix in my opinion

File: src/navier.f90 [hunk 2]
```diff
@@ -1211,12 +1210,12 @@ SUBROUTINE calc_varcoeff_rhs(pp3, rho1, px1, py1, pz1, dv3, drho1, ep1, divu3, r
...
-       CALL MPI_ALLREDUCE(rhomin,rho0,1,real_type,MPI_MIN,MPI_COMM_WORLD,ierr)
+       CALL MPI_ALLREDUCE(MPI_IN_PLACE,rhomin,1,real_type,MPI_MIN,MPI_COMM_WORLD,ierr)
     ENDIF
 
-    ta1(:,:,:) = (one - rho0 / rho1(:,:,:,1)) * px1(:,:,:)
-    tb1(:,:,:) = (one - rho0 / rho1(:,:,:,1)) * py1(:,:,:)
-    tc1(:,:,:) = (one - rho0 / rho1(:,:,:,1)) * pz1(:,:,:)
+    ta1(:,:,:) = (one - rhomin / rho1(:,:,:,1)) * px1(:,:,:)
+    tb1(:,:,:) = (one - rhomin / rho1(:,:,:,1)) * py1(:,:,:)
+    tc1(:,:,:) = (one - rhomin / rho1(:,:,:,1)) * pz1(:,:,:)
```
Comment (pbartholomew08) [inline src/navier.f90:1218]: For consistency with calling convention set `rho0=rhomin` after ALLREDUCE and use `rho0` as the numerator

**Review by pbartholomew08 — COMMENTED** [review]
I think this looks good

**Review by rfj82982 — APPROVED** [review]
LGTM

**Pull request merged**